### PR TITLE
[ValueTracking] Use `ConstantRange::toKnownBits` when computing from Range Metadata; NFC

### PR DESCRIPTION
**Pull request by goldsteinn**
Just replaces some bespoke logic with a better tested API.

### CODE REVIEW

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 1]
```diff
@@ -419,27 +419,16 @@ static void computeKnownBitsMul(const Value *Op0, const Value *Op1, bool NSW,
...
     ConstantRange Range(Lower->getValue(), Upper->getValue());
-
-    // The first CommonPrefixBits of all values in Range are equal.
-    unsigned CommonPrefixBits =
-        (Range.getUnsignedMax() ^ Range.getUnsignedMin()).countl_zero();
-    APInt Mask = APInt::getHighBitsSet(BitWidth, CommonPrefixBits);
-    APInt UnsignedMax = Range.getUnsignedMax().zextOrTrunc(BitWidth);
-    Known.One &= UnsignedMax & Mask;
-    Known.Zero &= ~UnsignedMax & Mask;
+    Known = Known.unionWith(Range.toKnownBits());
```
Comment (nikic) [inline llvm/lib/Analysis/ValueTracking.cpp:431]: Shouldn't this be intersectWith?

Comment (goldsteinn) [reply llvm/lib/Analysis/ValueTracking.cpp:431]: I think the precise logic would be an intersecting CR over all the ranges then union KnownBits with that final CR.
Ill update.

Comment (nikic) [reply llvm/lib/Analysis/ValueTracking.cpp:431]: Both are imprecise in different ways. Though afaik multi-range `!range` is never actually used, so it doesn't really matter.

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 2]
```diff
@@ -423,24 +423,17 @@ void llvm::computeKnownBitsFromRangeMetadata(const MDNode &Ranges,
...
     ConstantRange Range(Lower->getValue(), Upper->getValue());
-
-    // The first CommonPrefixBits of all values in Range are equal.
-    unsigned CommonPrefixBits =
-        (Range.getUnsignedMax() ^ Range.getUnsignedMin()).countl_zero();
-    APInt Mask = APInt::getHighBitsSet(BitWidth, CommonPrefixBits);
-    APInt UnsignedMax = Range.getUnsignedMax().zextOrTrunc(BitWidth);
-    Known.One &= UnsignedMax & Mask;
-    Known.Zero &= ~UnsignedMax & Mask;
+    CR = CR.intersectWith(Range.zextOrTrunc(BitWidth));
```
Comment (nikic) [inline llvm/lib/Analysis/ValueTracking.cpp:434]: This should be unionWith (see https://llvm.org/docs/LangRef.html#range-metadata). (And initialize to empty above.)

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 3]
```diff
@@ -423,24 +423,17 @@ void llvm::computeKnownBitsFromRangeMetadata(const MDNode &Ranges,
   unsigned NumRanges = Ranges.getNumOperands() / 2;
   assert(NumRanges >= 1);
 
-  Known.Zero.setAllBits();
-  Known.One.setAllBits();
+  ConstantRange CR = ConstantRange::getFull(BitWidth);
```
Comment (dtcxzyw) [inline llvm/lib/Analysis/ValueTracking.cpp:426]:
```suggestion
  ConstantRange CR = ConstantRange::getEmpty(BitWidth);
```

It should be initialized to empty :)

**Review by dtcxzyw — CHANGES_REQUESTED** [review]
> Failed Tests (40):
  LLVM :: Analysis/ScalarEvolution/shift-op.ll
  LLVM :: Analysis/ValueTracking/known-bits-from-range-md.ll
  LLVM :: CodeGen/AArch64/GlobalISel/prelegalizer-combiner-icmp-to-true-false-known-bits.mir
  LLVM :: CodeGen/AArch64/setcc_knownbits.ll
  LLVM :: CodeGen/AMDGPU/array-ptr-calc-i32.ll
  LLVM :: CodeGen/AMDGPU/divrem24-assume.ll
  LLVM :: CodeGen/AMDGPU/global-saddr-load.ll
  LLVM :: CodeGen/AMDGPU/load-range-metadata-sign-bits.ll
  LLVM :: CodeGen/AMDGPU/simple-indirect-call.ll
  LLVM :: CodeGen/AMDGPU/waterfall_kills_scc.ll
  LLVM :: CodeGen/PowerPC/BreakableToken-reduced.ll
  LLVM :: CodeGen/PowerPC/global-address-non-got-indirect-access.ll
  LLVM :: CodeGen/X86/pr12360.ll
  LLVM :: CodeGen/X86/pr48458.ll
  LLVM :: CodeGen/X86/pr48888.ll
  LLVM :: Transforms/GuardWidening/range-check-merging.ll
  LLVM :: Transforms/IndVarSimplify/shift-range-checks.ll
  LLVM :: Transforms/InferAddressSpaces/AMDGPU/ptrmask.ll
  LLVM :: Transforms/InstCombine/ARM/mve-v2i2v.ll
  LLVM :: Transforms/InstCombine/add2.ll
  LLVM :: Transforms/InstCombine/binop-cast.ll
  LLVM :: Transforms/InstCombine/bitreverse-known-bits.ll
  LLVM :: Transforms/InstCombine/gep-sext.ll
  LLVM :: Transforms/InstCombine/icmp-range.ll
  LLVM :: Transforms/InstCombine/icmp_sdiv_with_and_without_range.ll
  LLVM :: Transforms/InstCombine/ldexp.ll
  LLVM :: Transforms/InstCombine/narrow-math.ll
  LLVM :: Transforms/InstCombine/pr12251.ll
  LLVM :: Transforms/InstCombine/pr75129.ll
  LLVM :: Transforms/InstCombine/zext.ll
  LLVM-Unit :: Analysis/./AnalysisTests/ComputeKnownBitsTest/ComputeKnownBitsAddWithRange
  LLVM-Unit :: Analysis/./AnalysisTests/ComputeKnownBitsTest/ComputeKnownBitsAddWithRangeNoOverlap
  LLVM-Unit :: Analysis/./AnalysisTests/ComputeKnownBitsTest/ComputeKnownBitsGEPWithRange
  LLVM-Unit :: Analysis/./AnalysisTests/ComputeKnownBitsTest/ComputeKnownBitsGEPWithRangeNoOverlap
  LLVM-Unit :: Analysis/./AnalysisTests/ComputeKnownBitsTest/ComputeKnownBitsReturnedRangeConflict
  LLVM-Unit :: Analysis/./AnalysisTests/ComputeKnownBitsTest/KnownNonZeroShift
  LLVM-Unit :: CodeGen/./CodeGenTests/AArch64SelectionDAGTest/computeKnownBits_extload_known01
  LLVM-Unit :: CodeGen/./CodeGenTests/AArch64SelectionDAGTest/computeKnownBits_extload_knownnegative
  LLVM-Unit :: CodeGen/GlobalISel/./GlobalISelTests/AArch64GISelMITest/TestMetadata
  LLVM-Unit :: CodeGen/GlobalISel/./GlobalISelTests/AArch64GISelMITest/TestVectorMetadata

**Pull request closed without merging**